linux/mcompile/keymap/mc_import_rules.cpp

```diff
@@ -0,0 +1,651 @@
+    // Get the CAPSLOCK value
+   int capslock =
+        (this->KMX_IsCapsEqualToShift() ? 1 : 0) |
+        (this->KMX_IsSGCAPS() ? 2 : 0) |
+        (this->KMX_IsAltGrCapsEqualToAltGrShift() ? 4 : 0) |
+        (this->KMX_IsXxxxGrCapsEqualToXxxxShift() ? 8 : 0);
+
+    for (int ss = 0; ss <= MaxShiftState; ss++) {
+      if (ss == Menu || ss == ShftMenu) {
+        // Alt and Shift+Alt don't work, so skip them
+        continue;
+      }
+      for (int caps = 0; caps <= 1; caps++) {
+        std::u16string st = this->KMX_GetShiftState((ShiftState) ss, (caps == 1));
+
+        PKMX_WCHAR p;
+
+        if (st.size() == 0) {
+          // No character assigned here
+        }
+        else if (this->m_rgfDeadKey[(int)ss][caps]) {
+          // It's a dead key, append an @ sign.
+          key->dpContext = new KMX_WCHAR[1];
+          *key->dpContext = 0;
+
+          key->ShiftFlags = this->KMX_GetShiftStateValue(capslock, caps, (ShiftState) ss);
+          // we already use VK_US so no need to convert it as we do on windows
+          key->Key = this->VK();
+          key->Line = 0;
+
+          if(bDeadkeyConversion) {   // I4552
+            p = key->dpOutput = new KMX_WCHAR[2];
+            *p++ = st[0];
+            *p = 0;
+          } else {
+
+            p = key->dpOutput = new KMX_WCHAR[4];
+            *p++ = UC_SENTINEL;
+            *p++ = CODE_DEADKEY;
+            *p++ = KMX_DeadKeyMap(st[0], deadkeys, deadkeyBase, &KMX_FDeadkeys);   // I4353
+            *p = 0;
+          }
+          key++;
+        }
+        else {
+          bool isvalid = true;
+          for (size_t ich = 0; ich < st.size(); ich++) {
+            if(st[ich] < 0x20 || st[ich] == 0x7F) {
+              isvalid=false;
+              wprintf(L"invalid 16 for: %i\n", st[ich]);
+              break; }
+          }
+          if(isvalid) {
+            // this is different to mcompile windows !!!!
+            // this->m_sc    stores SC-US = SCUnderlying
+            // this->m_vk    stores VK-US ( not underlying !!)
+            // key->Key      stores VK-US ( not underlying !!)
+            // key->dpOutput stores character Underlying
+
+            KMX_DWORD SC_Underlying = KMX_get_KeyCodeUnderlying_From_KeyCodeUS(keymap, All_Vector, this->SC(), (ShiftState) ss, caps);
+            key->Key = KMX_get_VKUS_From_KeyCodeUnderlying( SC_Underlying);
+
+            key->Line = 0;
+            key->ShiftFlags = this->KMX_GetShiftStateValue(capslock, caps, (ShiftState) ss);
+
+            key->dpContext = new KMX_WCHAR;
+            *key->dpContext = 0;
+            p = key->dpOutput = new KMX_WCHAR[st.size() + 1];
+            for(size_t ich = 0; ich < st.size(); ich++) {
+              *p++ = st[ich];
+            }
+            *p = 0;
+            key++;
+          }
+        }
+      }
+    }
+    return true;
+  }
+};
+
+class KMX_Loader {
+private:
+  KMX_BYTE lpKeyStateNull[256];
```

Comment:
In the c'tor we only initialize it with 0s, but otherwise it's not used on Linux. On Windows it's used in a call to `ToUnicodeEx` in `ClearKeyboardBuffer`, but we don't have that on Linux, so I think this can safely go.